lib/src/tasks/coverage/api.dart

```diff
@@ -405,168 +438,157 @@ class CoverageTask extends Task {
+    return pr.stdout != null &&
+        (pr.stdout as String)
+            .contains(new RegExp(r'Error: Library not (found|supported)'));
+  }
 
-    String _testsFailedPattern = 'Some tests failed.';
-    String _testsPassedPattern = 'All tests passed!';
+  PubServeTask _startServeForCoverage() {
+    PubServeTask pubServeTask;
+    _coverageOutput.add('Starting Pub server...');
+
+    // Start `pub serve` on the `test` directory.
+    pubServeTask = startPubServe(
+        port: config.test.pubServePort, additionalArgs: ['test']);
+
+    _coverageOutput.add('::: ${pubServeTask.command}');
+    String indentLine(String line) => '    $line';
+
+    var startupLogFinished = new Completer();
+    pubServeTask.stdOut
+        .transform(until(startupLogFinished.future))
+        .map(indentLine)
+        .listen(_coverageOutput.add);
+    pubServeTask.stdErr
+        .transform(until(startupLogFinished.future))
+        .map(indentLine)
+        .listen(_coverageErrorOutput.add);
+
+    startupLogFinished.complete();
+    pubServeTask.stdOut.map(indentLine).join('\n').then((stdOut) {
+      if (stdOut.isNotEmpty) {
+        _coverageOutput
+            .add('`${pubServeTask.command}` (buffered stdout)');
+        _coverageOutput.add(stdOut);
+      }
+    });
+    pubServeTask.stdErr.map(indentLine).join('\n').then((stdErr) {
+      if (stdErr.isNotEmpty) {
+        _coverageOutput
+            .add('`${pubServeTask.command}` (buffered stdout)');
+        _coverageOutput.add(stdErr);
+      }
+    });
 
-    if (isBrowserTest) {
-      PubServeTask pubServeTask;
+    return pubServeTask;
+  }
 
-      try {
-        String testPath;
-        if (pubServe) {
-          _coverageOutput.add('Starting Pub server...');
-
-          // Start `pub serve` on the `test` directory.
-          pubServeTask = startPubServe(
-              port: config.test.pubServePort, additionalArgs: ['test']);
-
-          _coverageOutput.add('::: ${pubServeTask.command}');
-          String indentLine(String line) => '    $line';
-
-          var startupLogFinished = new Completer();
-          pubServeTask.stdOut
-              .transform(until(startupLogFinished.future))
-              .map(indentLine)
-              .listen(_coverageOutput.add);
-          pubServeTask.stdErr
-              .transform(until(startupLogFinished.future))
-              .map(indentLine)
-              .listen(_coverageErrorOutput.add);
-
-          PubServeInfo serveInfo = await pubServeTask.serveInfos.first;
-          if (!path.isWithin(serveInfo.directory, htmlFile.path)) {
-            throw '`pub serve` directory does not contain test file: ${htmlFile.path}';
-          }
+  Future<List<int>> _runTestOnVm(String testPath) async {
+    const String _observatoryFailPattern = 'Could not start Observatory HTTP server';
+    const String _testsFailedPattern = 'Some tests failed.';
+    const String _testsPassedPattern = 'All tests passed!';
```

Comment:
Since these constants are used both here and in `_runTestInBrowser()` below, we should move them out so that they're only defined once.